Call roles in typed fields

jakobandersen opened the pull request:
### Feature or Bugfix
- (Feature)
- Bugfix

### Purpose
As discussed in #8945, the typed field lists look like the use a specific domain role, but really they simply wrap the type in a ``pending_xref`` and set an attribute to tag with the role.
For some domains, e.g., C and C++, some types are not single xrefs, e.g., ``int*`` or ``std::pair<A, B>``, so they should really be parsed by the domain so it can typeset it correctly.
This PR looks up the role in the domain and calls it. The c, cpp, js, (and py ) domains are modified to use appropriate roles.

### Detail
- While it looks like the fields a processed as a post-transform, the transform is really called after parsing the directive content, so a docutils ``Inliner`` is sent into the transform for calling the role.
- For warnings it looks like the ``Inliner`` can not report the correct line, so the ``field`` node is sent into the transform as well for better location reporting.
- All affected function calls are simply extended with two extra parameters, similar to how the ``env`` was added.
- For Python there are mixin classes that further interprets the type such that ``a or b`` becomes multiple xrefs, so here the role calling is disabled by setting the ``Inliner`` to ``None``.
- If external domains have similar mixins they will break by this PR, but the fix is trivial (add the two parameters and pass them on). Though all these functions are undocumented so users shouldn't rely on this API.
- For v5 it may be good to clean up the field handling and simplify things.

### Relates
Fixes #8945.

jakobandersen committed:
js, use func role for exception field

The err role doesn't exist and the rest are all equivalent.

jakobandersen committed:
fields, call roles instead of making a pending_xref

Except for py which has mixins that assumes a single pending_xref

jakobandersen committed:
C, C++, use expr role in typed fields

jakobandersen committed:
Update CHANGES

tk0miya commented on sphinx/util/docfields.py:
```diff
@@ -62,39 +65,62 @@ def __init__(self, name: str, names: Tuple[str, ...] = (), label: str = None,
+        if role is None or inliner is None:
+            if role is None and inliner is not None:
+                msg = "Problem in %s domain: field is supposed "
+                msg += "to use role '%s', but that role is not in the domain."
```
Please use `__()` to make this message translatable.

tk0miya commented on sphinx/util/docfields.py:
```diff
@@ -62,39 +65,62 @@ def __init__(self, name: str, names: Tuple[str, ...] = (), label: str = None,
+            if inliner is None:
+                # msg = "Field for %s domain using role '%s' does not run the role."
+                # msg += " No inliner provided."
+                # logger.warning(msg, domain, rolename)
```
It seems this block is no longer needed, right?

tk0miya commented on sphinx/util/docfields.py:
```diff
@@ -62,39 +65,62 @@ def __init__(self, name: str, names: Tuple[str, ...] = (), label: str = None,
+            refnode = addnodes.pending_xref('', refdomain=domain, refexplicit=False,
+                                            reftype=rolename, reftarget=target)
+            refnode += contnode or innernode(target, target)
+            if env:
```
You don't need to ensure the existence because we assert it at the top of this method.

tk0miya commented on sphinx/util/docfields.py:
```diff
@@ -62,39 +65,62 @@ def __init__(self, name: str, names: Tuple[str, ...] = (), label: str = None,
-        if env:
-            env.get_domain(domain).process_field_xref(refnode)
-        return refnode
+        role = env.get_domain(domain).role(rolename)
```
Note (for me): the `domain` is passed from DocFieldTransformer. So it surely exists. So we don't need to take care the `env.get_domain()` raises an exception.

tk0miya reviewed (commented):
Sorry for late. LGTM with nits.

BTW, will you add testcases for this (or for C/C++ domains)?

tk0miya commented:
>For v5 it may be good to clean up the field handling and simplify things.

Agreed. I still don't understand what DocFieldTransformer does. It's very ugly and hard to understand.

jakobandersen commented:
> Sorry for late. LGTM with nits.
> 
> BTW, will you add testcases for this (or for C/C++ domains)?

Thanks for the review! Yes, I'll add test cases as well as addressing your comments.


> > For v5 it may be good to clean up the field handling and simplify things.
> 
> Agreed. I still don't understand what DocFieldTransformer does. It's very ugly and hard to understand.

To not forget, a note for a potential overhaul: I just remembered #3346 which seems relevant. I think going towards a mechanism where the domains gets more responsibility for  interpreting the field name/body might make something like the parameter direction easier to support.

jakobandersen committed:
DocFields, small fixes

jakobandersen committed:
Field roles, add C test

jakobandersen committed:
Field roles, add C++ test
